File: acwing/算法竞赛进阶指南/0x17-二叉堆/145. 超市.cpp
/*
维护一个最小堆，来一个新的商品，如果当前堆中商品个数等于 d_i，那么要放进去就必须拿出一个来，比较一下堆顶看值不值得放进去（p_i 大于堆顶的 p）即可。

y 总的思路更简单，不管怎样，先加进去，如果加进去的这个过期时间小于堆大小，然后 pop 一个出来，pop 出来的就是最小的了
*/
#include <algorithm>
#include <iostream>
using namespace std;
const int N = 10010;

int n, tot;
// 二叉堆模板，书上下标从 1 开始索引，这个例子是最小堆
pair<int, int> heap[N], d_p[N];
// 向上调整
void up(int p) {
    while (p > 1) {
        if (heap[p] < heap[p / 2]) {
            swap(heap[p], heap[p / 2]);
            p /= 2;
        } else
            break;
    }
}
void insert(pair<int, int> val) {
    heap[++tot] = val;
    up(tot);
}
// 向下调整
void down(int p) {
    // 左子节点
    int s = 2 * p;
    while (s <= tot) {
        // 左右子节点中取较小者
        if (s < tot && heap[s] > heap[s + 1]) s++;
        if (heap[p] > heap[s]) {
            swap(heap[p], heap[s]);
            p = s, s = 2 * p;
        } else
            break;
    }
}
int main() {
    while (cin >> n) {
        for (int i = 0; i < n; i++) {
            cin >> d_p[i].second >> d_p[i].first;
        }
        // 按照 d 从小到大排序
        sort(d_p, d_p + n);

        tot = 0;
        // 构建二叉堆
        for (int i = 0; i < n; i++) {
            int d = d_p[i].first, p = d_p[i].second;
            if (tot < d) {
                insert({p, d});
            } else if (tot == d) {
                if (heap[1].first < p) {
                    heap[1] = {p, d};
                    down(1);
                }
            } else
                break;
        }
        int sum = 0;
        // 求和
        for (int i = 1; i <= tot; i++) {
            sum += heap[i].first;
        }
        cout << sum << endl;
    }
}

File: acwing/算法竞赛进阶指南/0x17-二叉堆/145. 超市.js
class Heap {
  constructor (data = [], cmp = (a, b) => a < b) {
    if (typeof data === 'function') {
      cmp = data
      data = []
    }
    // heap 的数据一般从 1 开始索引，这样计算父亲、儿子节点更加简单
    this.data = [null]
    this.lt = (i, j) => cmp(this.data[i], this.data[j])
    data.forEach(this.push)
  }

  swap (i, j) {
    [this.data[i], this.data[j]] = [this.data[j], this.data[i]]
  }

  up (p) {
    while (p > 1) {
      if (this.lt(p, ~~(p / 2))) {
        this.swap(p, ~~(p / 2))
        p = ~~(p / 2)
      } else break
    }
  }

  down (p) {
    let s = p * 2
    while (s <= this.size()) {
      // 左右子节点取较小的（最小堆）/较大的（最大堆）
      if (s < this.size() && this.lt(s + 1, s)) s++
      // 子节点小于父亲节点，不满足最小堆性质
      if (this.lt(s, p)) {
        this.swap(s, p)
        p = s
        s = p * 2
      } else break
    }
  }

  remove (p) {
    this.swap(p, this.size())
    const val = this.data.pop()
    this.up(p)
    this.down(p)
    return val
  }

  size () {
    return this.data.length - 1
  }

  push (val) {
    this.data.push(val)
    this.up(this.size())
  }

  pop () {
    this.swap(1, this.size())
    const val = this.data.pop()
    this.down(1)
    return val
  }

  top () {
    return this.data[1]
  }
}

let input = ''
process.stdin.on('data', data => {
  input += data
})
process.stdin.on('end', () => {
  const arr = input.replace(/\n/g, ' ').split(' ').filter(s => s !== '').map(Number)
  // console.log(arr)
  // const iterator = arr[Symbol.iterator]()
  // console.log(arr)
  // console.log(iterator.next())
  let i = 0
  while (i < arr.length) {
    const n = arr[i++]
    const heap = new Heap()
    const pd = []
    for (let j = 0; j < n; j++) {
      pd.push([arr[i++], arr[i++]])
    }
    // 按照过期时间排序，保证插入到堆中的时候，堆中已经存在的商品的过期时间不会大于当前商品的过期时间
    pd.sort((a, b) => a[1] - b[1])
    for (let j = 0; j < n; j++) {
      const [p, d] = [pd[j][0], pd[j][1]]
      if (d > heap.size()) {
        heap.push(p)
      } else if (d === heap.size()) {
        if (heap.top() < p) {
          heap.pop()
          heap.push(p)
        }
      }
      // console.log(p, d, heap.data)
    }
    console.log(heap.data.slice(1).reduce((p, c) => p + c, 0))
  }
})
